{
    "id": "SOLR-2975",
    "title": "Solr test failure when running under Java 1.5",
    "details": {
        "affect_versions": "3.5,                                            3.6",
        "status": "Closed",
        "fix_versions": [
            "3.6",
            "4.0-ALPHA"
        ],
        "components": [
            "contrib - DataImportHandler"
        ],
        "type": "Test",
        "priority": "Minor",
        "labels": "",
        "resolution": "Fixed"
    },
    "description": "ant test -Dtestcase=TestSolrEntityProcessorUnit fails when running under Java 1.5 because of faulty assumptions in the test.\n\nFrom e-mail thread (Hossman):\n...those lines are assuming that row.entrySet will return something that\nhas a predictible iteration order, but row is a Map of unknown creation\n(returned by the entityProcessor) ... so unless the entityProcessor is\nexplicitly defined as returning something like SortedMap (which isn't\nsuggested anywhere in this test) the test is making a really bad\nassumption.\n\nFrom e-mail. (Steven Rowe)\nFYI, I see this same failure when I run the branch_3x tests with Java 1.5, but not 1.6.\nand\nOh, and the reason Jenkins isn't seeing this failure is that it runs branch_3x tests using Java 1.6, after first compiling with Java 1.5\n\nEven though we won't run Solr 4 under java 1.5, I'll change it there anyway since this is a bad assumption in the test.",
    "attachments": {
        "SOLR-2975.patch": "https://issues.apache.org/jira/secure/attachment/12507698/SOLR-2975.patch"
    },
    "issue_links": {},
    "comments": [
        {
            "author": "Erick Erickson",
            "id": "comment-13171004",
            "date": "2011-12-16T14:59:30+0000",
            "content": "Running full test now, will check in shortly unless someone objects.\n\nApplies to both trunk and 3x "
        },
        {
            "author": "Martijn van Groningen",
            "id": "comment-13171078",
            "date": "2011-12-16T17:11:46+0000",
            "content": "Attached refactored version of test. I ran the test 10000 times on my machine with Java 1.5 and didn't have failures.\n\nThere was also a concurrency issue in the testMultiThread() method. The access to the rowList variable wasn't synchronized. Also the ThreadPoolExecutor wasn't cleaned up. This resulted in memory issues and thread issues (OS couldn't create threads) if the tests were run for more then 2000 times. "
        },
        {
            "author": "Erick Erickson",
            "id": "comment-13171105",
            "date": "2011-12-16T17:55:37+0000",
            "content": "Fixes my failure up too. When do you want to commit this? I can take care of it if you want.\n\nThanks lots, now I know about SolrTestCaseJ4.Doc, which solves the problem I was having when I tried to use a List<Map<String, Object>> and couldn't get the multiValued test to work..... "
        },
        {
            "author": "Martijn van Groningen",
            "id": "comment-13171188",
            "date": "2011-12-16T20:54:43+0000",
            "content": "If you want you can commit the fix to svn. You already assigned the issue  "
        },
        {
            "author": "Erick Erickson",
            "id": "comment-13171574",
            "date": "2011-12-17T13:46:53+0000",
            "content": "trunk - r1215487\n3x    - r1215486\n\nThanks Martijn! "
        }
    ]
}